package/samplers/cmamae/sampler.py

```diff
@@ -0,0 +1,191 @@
+            ranges=archive_ranges,
+        )
+        emitters = [
+            EvolutionStrategyEmitter(
+                archive,
+                x0=emitter_x0_np,
+                sigma0=emitter_sigma0,
+                ranker="imp",
+                selection_rule="mu",
+                restart_rule="basic",
+                batch_size=emitter_batch_size,
+            ) for _ in range(n_emitters)
+        ]
+
+        # Number of solutions generated in each batch from pyribs.
+        self._batch_size = n_emitters * emitter_batch_size
+
+        self._scheduler = Scheduler(
+            archive,
+            emitters,
+            result_archive=result_archive,
+        )
+
+        self._values_to_tell: list[list[float]] = []
+
+    def _validate_params(self, param_names: list[str],
+                         emitter_x0: dict[str, float]) -> None:
+        dim = len(param_names)
+        param_set = set(param_names)
+        if dim != len(param_set):
+            raise ValueError(
+                "Some elements in param_names are duplicated. Please make it a unique list."
+            )
+
+        if set(param_names) != emitter_x0.keys():
+            raise ValueError(
+                "emitter_x0 does not contain the parameters listed in param_names. "
+                "Please provide an initial value for each parameter.")
+
+    def _validate_param_names(self, given_param_names: Iterable[str]) -> None:
+        if set(self._param_names) != set(given_param_names):
+            raise ValueError("The given param names must match the param names "
+                             "initially passed to this sampler.")
+
+    def _convert_to_pyribs_params(self, params: dict[str, float]) -> np.ndarray:
+        np_params = np.empty(len(self._param_names), dtype=float)
+        for i, p in enumerate(self._param_names):
+            np_params[i] = params[p]
+        return np_params
+
+    def _convert_to_optuna_params(self, params: np.ndarray) -> dict[str, float]:
+        dict_params = {}
+        for i, p in enumerate(self._param_names):
+            dict_params[p] = params[i]
+        return dict_params
+
+    def sample_relative(
+            self, study: Study, trial: FrozenTrial,
+            search_space: dict[str, BaseDistribution]) -> dict[str, float]:
+        self._validate_param_names(search_space.keys())
+
+        # Note: Batch optimization means we need to enqueue trials.
+        solutions = self._scheduler.ask()
+        next_params = self._convert_to_optuna_params(solutions[0])
+        for solution in solutions[1:]:
+            params = self._convert_to_optuna_params(solution)
+            study.enqueue_trial(params)
+
+        return next_params
+
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Sequence[float] | None,
+    ) -> None:
+        # TODO: Is it safe to assume the parameters will always come back in the
```

Comment:
Is it safe to assume the parameters will always come back in the order that they were sent out by the scheduler / sample_relative function? For example, if I send out solutions A, B, C, D, will they come back as A, B, C, D, or will they come back in some other order like D, B, C, A?

Pyribs makes the assumption that things are in order. If this is not the case, maybe we can retrieve solutions based on their trial ID?